game-core/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java

```diff
@@ -148,7 +148,7 @@ public void unitsLostInPrecedingBattle(
     if (withdrawn) {
       return;
     }
-    Collection<Unit> lost = getDependentUnits(units);
+    Collection<Unit> lost = new ArrayList<>(getDependentUnits(units));
```

Comment:
I agree this is something we should make a decision on, but I propose as a rule of thumb getters should always return immutable collections.

It has the following benefits:
  1. Just accessing the collection read-only does not incur performance cost of copying it.
  2. Modifying the return value becomes never acceptable. When it is acceptable, like with the current state, now one has to go look at the implementation to make sure modifying it doesn't change the underlying copy.
  3. And the last benefit, which is over a plain accessor that's neither immutable nor a copy, is the caller can't go around modifying the inner state of some other class after calling a getter. Returning a copy also avoids this issue.

Now, I agree that in the case of getDependentUnits(), returning an immutable view actually ends up incurring a performance cost if the caller wants to start modifying it, but there's value in consistency. If some methods return mutable copies and others immutable views, then it's very hard to reason about.

Another option that could satisfy things would be to adopt different naming conventions. For example getDependentUnitsCopy() - but this still incurs more mental complexity than just treating each accessor as providing an immutable view and having the caller be responsible for copying it, if it needs to.

Reply:
Appreciate the discussion, some quick thoughts:
- modifiable vs mutable collection are not quite the same. Just because a list is unmodifiable does not mean its contents are immutable. We probably should be doing deep copies of object state if we are really going to be going for immutability, and/or just use immutable objects in umodifiable lists

> Just accessing the collection read-only does not incur performance cost of copying it.

Have you measured the performance hit?

> Modifying the return value becomes never acceptable. When it is acceptable, like with the current state, now one has to go look at the implementation to make sure modifying it doesn't change the underlying copy.

The problem is we are working with a legacy code base, it's likely quite common to have this situation. We should try to survey how much we have of each case. It could be the case that there is so much modification of returned lists, that it's not very tractable to ever really fix it all. If so, we'll at best be in a half and half state and not benefit from any convention.

-----------

In part we are debating whether it should the caller or callee that makes a copy to avoid mutating an underlying collection. We should also not overlook the collection contents as well.

Without numbers, given the existing code is likely to favor copying at the callee side, IMO that is probably the safer approach. The unmodifiable collection code is more recent, there is likely less of it. We should also know the performance penalty we are actually talking about as well. @asvitkine would you be able to measure that? (FYI there is a [PerfTimer](https://github.com/triplea-game/triplea/blob/master/game-core/src/main/java/org/triplea/performance/PerfTimer.java) for convenience to help with such measurements)

Reply:
> Appreciate the discussion, some quick thoughts:
> 
> * modifiable vs mutable collection are not quite the same. Just because a list is unmodifiable does not mean its contents are immutable. We probably should be doing deep copies of object state if we are really going to be going for immutability, and/or just use immutable objects in umodifiable lists
> 
> > Just accessing the collection read-only does not incur performance cost of copying it.
> 
> Have you measured the performance hit?

I have not, but obviously it depends how often this is called and how big the collections are. I suspect returning a modifiable collection is essentially free, so we can make a silly test of doing a loop N times for some large N (e.g. 1 mil) and calling the version that copies and the version that doesn't and see some ridiculous speedup when testing those two in isolation (like thousands of times faster). But obviously, that's not a real world example.

Measuring within the context of the game, on the other hand, is hard - since I don't think any particular operation is significantly slow - but the aggregate might be non trivial - but that's hard to measure, because now we're also measuring all the other parts of the game engine and I suspect there might be a lot of variance there from other parts. So it might be hard to get a clear signal if there's a lot of variance.

> 
> > Modifying the return value becomes never acceptable. When it is acceptable, like with the current state, now one has to go look at the implementation to make sure modifying it doesn't change the underlying copy.
> 
> The problem is we are working with a legacy code base, it's likely quite common to have this situation. We should try to survey how much we have of each case. It could be the case that there is so much modification of returned lists, that it's not very tractable to ever really fix it all. If so, we'll at best be in a half and half state and not benefit from any convention.
> 
> In part we are debating whether it should the caller or callee that makes a copy to avoid mutating an underlying collection. We should also not overlook the collection contents as well.
> 
> Without numbers, given the existing code is likely to favor copying at the callee side, IMO that is probably the safer approach. The unmodifiable collection code is more recent, there is likely less of it. We should also know the performance penalty we are actually talking about as well. @asvitkine would you be able to measure that? (FYI there is a [PerfTimer](https://github.com/triplea-game/triplea/blob/master/game-core/src/main/java/org/triplea/performance/PerfTimer.java) for convenience to help with such measurements)

I'm likely not going to be able to measure it soon as I'm out of town this weekend. But really, this change should be a clear win.

...

Also one other thing I forgot to mention that's the advantage of consistently switching to this pattern - we will prevent double copying. In the current code, you often see both the getter doing a copy and the caller doing another copy, since the caller is not sure if it's safe to edit the returned one. If we switch to this consistently, then there will be at most 1 copy and in cases where no modification are required, 0 copies. So it should be a clear reduction in work done.